packages/babel-types/src/modifications/typescript/removeTypeDuplicates.js

```diff
@@ -0,0 +1,74 @@
+import {
+  isTSAnyKeyword,
+  isTSUnionType,
+  isTSTypeParameter,
+} from "../../validators/generated";
+
+/**
+ * Dedupe type annotations.
+ */
+export default function removeTypeDuplicates(
+  nodes: Array<Object>,
+): Array<Object> {
+  const generics = {};
+  // FIXME: I don't understand the purpose of bases for Flow (if it has an analogue for TS)
+  // const bases = {};
+
+  // store union type groups to circular references
+  const typeGroups = [];
+
+  const types = [];
+
+  for (let i = 0; i < nodes.length; i++) {
+    const node = nodes[i];
+    if (!node) continue;
+
+    // detect duplicates
+    if (types.indexOf(node) >= 0) {
+      continue;
+    }
+
+    if (isTSAnyKeyword(node)) {
+      return [node];
+    }
+
+    // FIXME: I don't understand the purpose of bases for Flow (if it has an analogue for TS)
+    // if (isFlowBaseAnnotation(node)) {
+    //   bases[node.type] = node;
+    //   continue;
+    // }
+
+    if (isTSUnionType(node)) {
+      if (typeGroups.indexOf(node.types) < 0) {
+        nodes = nodes.concat(node.types);
+        typeGroups.push(node.types);
+      }
+      continue;
+    }
+
+    if (isTSTypeParameter(node)) {
```

Comment:
As I understand `TSTypeParameter` (in comparison to `GenericTypeAnnotation`) doesn't have parameters to merge as it works for Flow
example: https://github.com/babel/babel/blob/master/packages/babel-parser/test/fixtures/typescript/types/conditional-infer/output.json#L23

Reply:
❌ Removed to clarify how exactly I need to replace `GenericTypeAnnotation`